fontforge/splineutil2.c

```diff
@@ -2879,6 +2880,159 @@ void SplineCharAddExtrema(SplineChar *sc, SplineSet *head,enum ae_type between_s
 	    SplineSetAddExtrema(sc,ss,between_selected,emsize);
 }
 
+Spline *SplineAddInflections(Spline *s) { 
+    if ( s->knownlinear )
+return(s);
+    /* First find the inflections, if any */
+    extended inflect[2];
+    int i = Spline2DFindPointsOfInflection(s, inflect);
+    if ( i==2 ) {
+	if ( RealNearish(inflect[0],inflect[1]) )
+	    --i;
+	else if ( inflect[0]>inflect[1] ) {
+	    real temp = inflect[0];
+	    inflect[0] = inflect[1];
+	    inflect[1] = temp;
+	}
+    }   
+    /* Now split the spline at 1 or 2 times */
+    extended splittimes[3] = {-1,-1,-1}; /* 3 slots due to SplineSplit() */
+    for (int j = 0; j < i; j++) {
+		if ( inflect[j] > .001 && inflect[j] < .999 ) /* avoid rounding issues */
+			splittimes[j] = inflect[j];
+	}
+    if ( splittimes[0] == -1)
+return(s);
+    s = SplineSplit(s,splittimes);
+return(s);	
+}
+
+void SplineSetAddInflections(SplineChar *sc, SplineSet *ss, int anysel) {
+    Spline *s, *first; 
+    first = NULL;
+    for ( s = ss->first->next; s!=NULL && s!=first; s = s->to->next ) {
+	    if ( !anysel || s->from->selected && s->to->selected )
+			s = SplineAddInflections(s);
+	    if ( first==NULL ) first = s;
+    }
+}
+
+void SplineCharAddInflections(SplineChar *sc, SplineSet *head, int anysel) { 
+    SplineSet *ss;
+    for ( ss=head; ss!=NULL; ss=ss->next )
+	    SplineSetAddInflections(sc,ss,anysel);
+}
+
+/* Make the line between the control points parallel to the chord */
+/* such that the area is preserved (kind of an improved "tunnify") */
+Spline *SplineBalance(Spline *s) { 
```

Comment:
You're (obviously -- this is the point) modifying SplinePoint values in this function but I don't see calls to `SplineRefigure()` anywhere in the path. (Above that's taken care of by SplineSplit().)

Reply:
Indeed, `SplineRefigure()` is not called ( I just have checked it) although it probably should be called. I HAVE tested my PR and I did not run into problems so far. But my tests were not more systematic than testing some problematic cases and run the functions on some huge fonts. I did not implement a separate torture test for the UI functions. Is there an existing torture test that I have missed (like trip and trap by Donald E. Knuth)?
Anyway, I will add `SplineRefigure()` calls as soon as I have a little bit more time.

Reply:
There is absolutely no good systematic method of UI testing, unfortunately. When one is adding or making significant modifications to UI-tied functionality it's best to just play around with it in the UI to see what happens. Making real-time changes in a CharView is probably the best way as its most likely to turn up updating problems (such as those caused by a lack of `SplineRefigure()`.